docs/04-smart-contracts/13-testing.md

```diff
@@ -0,0 +1,269 @@
+---
+sidebar_position: 13
+---
+
+# Testing
+
+At some point in Smart Contract development, everyone will arrive at tests. Tests are what give you and your protocol sleep at night and ensure changes can be rapidly deployed to the contracts codebase without breaking everything else.
+
+A great set of contracts will have a great set of tests generally divided into two areas of testing. Unit testing and Integration testing. 
+
+
+## Unit Testing
+
+See here for a [guide on unit testing](https://docs.cosmwasm.com/tutorials/simple-option/testing)
+
+## Integration Testing with `cw-multi-test`
+
+The cw-multi-test package offered in the cw-plus repo provides an interesting way to test your smart contracts without going all the way to deploying them on a testnet. Before using multi-test the flow to me was to have some pipeline which would set up your contracts on a given chain (maybe testnet, maybe local) perform some tests and then if possible destroy/self-destruct the contracts.
+
+All of that can be taken away almost in preference of cw-multi-test-based integration tests which enable you to test the flows and interactions between smart contracts. There is still a place for the flow described above but I have had a better experience writing these integration tests once you figure out the intricacies of multi-test. I hope to clear some of those intricacies up here with some tips, resources and steps.
+
+## `cw-multi-test` concepts
+
+There are a few main concepts to understand before you will be able to  simulate a blockchain environment in Rust and run tests that involve contract -> contract,
+and contract -> bank interactions.
+
+In this section we will take a step-by-step look through writing a test with cw-multi-test, explaining some important concepts along the way. 
+To start we need a specimen contract such as the [cw-template](https://github.com/InterWasm/cw-template/blob/main/src/contract.rs) which is a simple boilerplate contract containing two functions: `Increment` and `Reset`. While there are already unit tests for the cw-template, there is no `cw-multi-test` tests to simulate an entire run through. 
```

Comment:
Update: integration tests has been added to the cw-template https://github.com/InterWasm/cw-template/pull/105
```suggestion
To start we need a specimen contract such as the [cw-template](https://github.com/InterWasm/cw-template/blob/main/src/contract.rs) which is a simple boilerplate contract containing two functions: `Increment` and `Reset`.
```